src/library_html5.js

```diff
@@ -2051,30 +2051,27 @@ var LibraryJSEvents = {
 #if ASSERTIONS
       assert(e);
 #endif
-      var touches = {};
-      var et = e.touches;
+      var t, touches = {}, et = e.touches;
+      // To ease marshalling different kinds of touches that browser reports (all touches are listed in e.touches, 
+      // only changed touches in e.changedTouches, and touches on target at a.targetTouches), mark a boolean in
+      // each Touch object so that we can later loop only once over all touches we see to marshall over to Wasm.
+
       for (var i = 0; i < et.length; ++i) {
-        var touch = et[i];
-#if ASSERTIONS
-        // Verify that browser does not recycle touch objects with old stale data, but reports new ones each time.
-        assert(!touch.isChanged);
-        assert(!touch.onTarget);
-#endif
-        touches[touch.identifier] = touch;
+        t = et[i];
+        // Browser might recycle the generated Touch objects between each frame (Firefox on Android), so reset any
+        // changed/target states we may have set from previous frame.
+        t.isChanged = t.onTarget = 0;
+        touches[t.identifier] = t;
       }
-      et = e.changedTouches;
-      for (var i = 0; i < et.length; ++i) {
-        var touch = et[i];
-#if ASSERTIONS
-        // Verify that browser does not recycle touch objects with old stale data, but reports new ones each time.
-        assert(!touch.onTarget);
-#endif
-        touch.isChanged = 1;
-        touches[touch.identifier] = touch;
+      // Mark which touches are part of the changedTouches list.
+      for (var i = 0; i < e.changedTouches.length; ++i) {
+        t = e.changedTouches[i];
+        t.isChanged = 1;
+        touches[t.identifier] = t;
       }
-      et = e.targetTouches;
-      for (var i = 0; i < et.length; ++i) {
-        touches[et[i].identifier].onTarget = 1;
+      // Mark which touches are part of the targetTouches list.
+      for (var i = 0; i < e.targetTouches.length; ++i) {
```

Comment:
Is it more compact to write `for (var i in e.targetTouches)` here and above?

Reply:
I think it would need to be `of` and not `in`, as it is an array. That would be shorter, but is not supported in IE and some other older browsers. But maybe at some point we should say we don't care about them?

Reply:
This was resolved in https://github.com/emscripten-core/emscripten/pull/15094